samples/msal-node-samples/nextjs-rsc-ts/src/utils/AuthProvider.ts

```diff
@@ -0,0 +1,285 @@
+      },
+      payload
+    );
+
+    return {
+      account: authResult.account,
+      returnTo: state.returnTo,
+    };
+  }
+
+  /**
+   * Authenticate a user.
+   * @returns The logged in account along with an instance that is configured with a partitioned cache.
+   * @remarks Can safely be called in multiple server components.
+   */
+  authenticate = cache(async () => {
+    const partitionManager = await this.partitionManagerFactory();
+    const homeAccountId = await partitionManager.getKey();
+
+    const instance = await this.getInstance();
+
+    const account = homeAccountId
+      ? await instance.getTokenCache().getAccountByHomeId(homeAccountId)
+      : null;
+
+    return { account, instance };
+  });
+
+  /**
+   * Get the current logged in account.
+   * @returns An account object if a user is logged in, or null if no user is logged in.
+   * @remarks Can safely be called in multiple server components.
+   * @remarks Prefer authenticate() in Server Actions and Route Handlers
+   */
+  getAccount = cache(async () => {
+    const partitionManager = await this.partitionManagerFactory();
+    const homeAccountId = await partitionManager.getKey();
+
+    if (!homeAccountId) {
+      return null;
+    }
+
+    const instance = await this.getInstance();
+
+    return await instance.getTokenCache().getAccountByHomeId(homeAccountId);
+  });
+
+  /**
+   * Get an instance configured with a partitioned cache to the current logged in user.
+   * @returns A ConfidentialClientApplication instance
+   * @remarks Can safely be called in multiple server components.
+   * @remarks Prefer authenticate() in Server Actions and Route Handlers
+   */
+  getInstance = cache(async () => {
+    const cachePlugin = new DistributedCachePlugin(
+      this.cacheClient,
+      await this.partitionManagerFactory()
+    );
+
+    const config = {
+      ...this.configuration,
+      cache: {
+        cachePlugin,
+      },
+    };
+
+    if (!config.auth.cloudDiscoveryMetadata || !config.auth.authorityMetadata) {
+      const metadata = await this.getMetadata(
+        config.auth.clientId,
+        config.auth.authority ?? "https://login.microsoftonline.com/common"
+      );
+
+      if (metadata) {
+        config.auth.cloudDiscoveryMetadata = metadata.cloudDiscoveryMetadata;
+        config.auth.authorityMetadata = metadata.authorityMetadata;
+      }
+    }
+
+    return new ConfidentialClientApplication(config);
+  });
+
+  // validate that the payload includes required fields
+  private getAuthorizationCodePayload(formData: FormData) {
+    // validate that we only get string entries
+    const stringEntries = Array.from(formData.entries()).filter(
+      ([, value]) => typeof value === "string"
+    );
+
+    const data = Object.fromEntries(stringEntries);
+
+    if (!("state" in data)) {
+      throw new Error("No state found in payload.");
+    }
+
+    if (!("code" in data)) {
+      throw new Error("No code found in payload.");
+    }
```

Comment:
State is needed before it is passed to MSAL, as it stores the request made, which is needed for incremental consent. I'm not sure it's the right place to store it though.

As for code, the types of msal doesn't allow me to pass a null/undefined `code` value, so based on the types alone I wouldn't think msal could throw its own errors even. Since I cast it to the required type below, I want to validate that the required properties are present.

Reply:
Got it, thanks for elaborating this. About storing the initial request via the state param -If possible, I would prefer using the session for this, since the state param travels over the address bar i.e. goes into the browser history. Not that there's any sensitive info in the request object here, but more of a general precaution.

Reply:
I need to think about this, because of the way I made it right now. The `AuthProvider` doesn't really do any session-specific logic, or nextjs specific logic, other that calling the partition manager. Currently I also use session with cookie storage, and should probably switch to something else due to cookie size limitations.

I'm thinking of the following options for now;

- Use [`createSessionStorage`](https://remix.run/docs/en/1.17.1/utils/sessions#createsessionstorage) with redis, but again that's from remix which I should probably move away from in this sample.
- Use the next version of [`iron-session`](https://github.com/vvo/iron-session/issues/586), which is framework-agnostic and would support app router, however this is in alpha.
- Use `AuthProvider.cacheClient` to store the request, with a key generated from `AuthProvider.cryptoProvider.createNewGuid()`, which i then store in the state, or session cookie storage.

Reply:
I see. In that case I would say using the state for storing request object is fine, as far as samples go. I'd feel easier though if we could add some comments to encourage using sessions whenever possible, or if they have to use the state param, make sure to not include sensitive information in it.

Reply:
@h3rmanj where may I find the alpha release that supports app router? What is your solution to work with app router + iron-session in the short-medium development term?

Reply:
@renchris I haven't tried integrating with it yet. The alpha is released under the `alpha` tag on npm, but no docs are available yet. The issue I linked has some discussion with people managing to integrate it.

Reply:
Added a remark in the README.